core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/state/ReplicationState.java

```diff
@@ -41,16 +44,47 @@ public ReplicationState(Metadata metadata, Host localhost)
 
     public Map<LongTokenRange, ImmutableSet<Host>> getTokenRangeToReplicas(TableReference tableReference)
     {
+        String keyspace = tableReference.getKeyspace();
+
         Map<LongTokenRange, ImmutableSet<Host>> tokenRangeToReplicaMap = new HashMap<>();
 
-        for (TokenRange tokenRange : myMetadata.getTokenRanges(tableReference.getKeyspace(), myLocalHost))
+        Map<ImmutableSet<Host>, ImmutableSet<Host>> replicaCache = new HashMap<>();
+
+        for (TokenRange tokenRange : myMetadata.getTokenRanges(keyspace, myLocalHost))
         {
             LongTokenRange longTokenRange = convert(tokenRange);
-            ImmutableSet<Host> hosts = ImmutableSet.copyOf(myMetadata.getReplicas(tableReference.getKeyspace(), tokenRange));
-            tokenRangeToReplicaMap.put(longTokenRange, hosts);
+            ImmutableSet<Host> replicas = ImmutableSet.copyOf(myMetadata.getReplicas(keyspace, tokenRange));
+
+            tokenRangeToReplicaMap.put(longTokenRange, getOrCache(replicas, replicaCache));
+        }
+
+        return getOrCache(keyspace, ImmutableMap.copyOf(tokenRangeToReplicaMap));
+    }
+
+    private ImmutableSet<Host> getOrCache(ImmutableSet<Host> replicas, Map<ImmutableSet<Host>, ImmutableSet<Host>> replicaCache)
+    {
+        ImmutableSet<Host> actualReplicas;
+        actualReplicas = replicaCache.putIfAbsent(replicas, replicas);
+        if (actualReplicas == null)
+        {
+            actualReplicas = replicas;
+        }
+
+        return actualReplicas;
+    }
+
+    private ImmutableMap<LongTokenRange, ImmutableSet<Host>> getOrCache(String keyspace, ImmutableMap<LongTokenRange, ImmutableSet<Host>> replication)
```

Comment:
Should this method be synchronized? If not, then I suppose that keyspaceReplicationCache could be an ordinary HashMap<>.

Reply:
Removed it and used ConcurrentHashMap#compute instead